### PR TITLE
add CharacterEncodingFilter to spring security chain

### PR DESCRIPTION
**Pull request by anzioka**

### CODE REVIEW

File: src/main/java/com/openlattice/chronicle/pods/ChronicleServerSecurityPod.java [hunk 1]
```diff
@@ -43,6 +47,10 @@ protected void authorizeRequests( HttpSecurity http ) throws Exception {
                 .antMatchers( HttpMethod.OPTIONS ).permitAll()
                 .antMatchers( HttpMethod.POST, "/chronicle/data/study/participant" ).permitAll()
                 .antMatchers ( "/chronicle/data/authenticated/**" ).authenticated();
-    }
 
+        CharacterEncodingFilter filter = new CharacterEncodingFilter();
+        filter.setEncoding( StandardCharsets.UTF_8.toString() );
+        filter.setForceEncoding( true );
+        http.addFilterBefore( filter, CsrfFilter.class );
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/pods/ChronicleServerSecurityPod.java:54]: How do you know to do this? Why before the `CsrfFilter` class specifically?

Comment (anzioka) [reply src/main/java/com/openlattice/chronicle/pods/ChronicleServerSecurityPod.java:54]: After researching, it seems that the recommended solution is wrapping a CharacterEncodingFilter in a  FilterRegistrationBean and making sure that it is first in the filter chain to handle requests. So I added a bean to the servlets pod and ordered it first:
```
    @Bean (name = "encodingFilter")
    @Order ( Ordered.HIGHEST_PRECEDENCE )
    public FilterRegistrationBean<CharacterEncodingFilter> characterEncodingFilter() {
        CharacterEncodingFilter filter = new CharacterEncodingFilter();
        filter.setEncoding( "UTF-8" );
        filter.setForceEncoding( true );

        FilterRegistrationBean<CharacterEncodingFilter> registrationBean = new FilterRegistrationBean<>();
        registrationBean.setFilter( filter );
        registrationBean.addUrlPatterns( "/*" );

        return registrationBean;
    }
```

Unfortunately, this does not work although `applicationContext.getBeanDefinitionNames()` confirmed that the bean was added.

Next I tried adding the filter to the default fallback security chain (above solution) and it worked as expected. I have updated the code to explicitly ensure that it is ordered first in the chain:

2021-07-15T07:51:28,740 INFO [main] org.springframework.security.web.DefaultSecurityFilterChain - Creating filter chain: any request, [org.springframework.web.filter.CharacterEncodingFilter@4de93edd, org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter@53b2e1eb, org.springframework.security.web.context.SecurityContextPersistenceFilter@1b7a4930, org.springframework.security.web.header.HeaderWriterFilter@5f01fb5c, org.springframework.security.web.authentication.logout.LogoutFilter@10d49900, org.springframework.security.web.savedrequest.RequestCacheAwareFilter@4bfe83d, org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter@2a504ea7, org.springframework.security.web.authentication.AnonymousAuthenticationFilter@29db008c, org.springframework.security.web.session.SessionManagementFilter@240f2efd, org.springframework.security.web.access.ExceptionTranslationFilter@3991fe6d, org.springframework.security.web.access.intercept.FilterSecurityInterceptor@434ee422]

However, placing it after FilterSecurityInterceptor also works as expected, so I don't think it matters.

**Review by UnsungHero97 — DISMISSED** [review]

File: src/main/java/com/openlattice/chronicle/pods/ChronicleServerSecurityPod.java (hunk 1 above)
Comment (anzioka) [reply src/main/java/com/openlattice/chronicle/pods/ChronicleServerSecurityPod.java]: Reference:
https://stackoverflow.com/a/23051264
https://www.baeldung.com/spring-boot-characterencodingfilter

**Review by UnsungHero97 — APPROVED** [review]

**Review by geekbeast — APPROVED** [review]
Doesn't seem like the ideal place to do it, but good enough for now.

**Pull request merged**